### PR TITLE
[Opt] Merge a sequence of mat muls into one large mat mul.

### PR DESCRIPTION
**Pull request by nadavrot**
Optimize a sequence of matrix multipliers into one large matrix multiplication. Larger matrix multiplications are more likely to utilize the hardware because they expose more data-level parallelism.  Notice that this commit only implements the combination of LHS matrices. Combining RHS matrices will land in a future commit.

### CODE REVIEW

File: lib/Optimizer/GraphOptimizer.cpp [hunk 1]
```diff
@@ -422,6 +422,121 @@ static bool sinkCode(Function *F) {
...
+// M| A  |  T| B  | * K|    C    | =    |---------|
+//   ---- ,  |    |    |         |     T|  B * C  |
+//    K       ----      ---------        ---------
+//             K            R                R
+static void mergeMatMul(Function *F) {
+  auto &nodes = F->getNodes();
+
+  // These two maps record the list of matrix multipliers that use each node
+  // value either as a right-hand-side user or a left-hand-user.
+  llvm::DenseMap<Node *, std::vector<MatMulNode *>> righMatrixUsers;
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer.cpp]: `righMatrixUsers` -> `rightMatrixUsers` ?

File: lib/Optimizer/GraphOptimizer.cpp [hunk 2]
```diff
@@ -422,6 +422,121 @@ static bool sinkCode(Function *F) {
...
+    // Collects the LHS values to merge.
+    std::vector<NodeValue> LHS;
+
+    // For each matmul that depends on the rhs matrix.
+    for (auto &MM : MMs) {
+      auto L = MM->getLHS();
+      // The operands to the matrix multiplier should not depend on one another
+      // or else we won't be able to get rid of the original matrix
+      // multiplication.
+      if (mayDependOnAny(LHS, L.getNode())) {
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer.cpp]: If the default depth of 6 for `mayDepend()` is too small and a dependency does exist, would an assert somewhere later on catch this?

**Comment by nadavrot** [comment]
@jfix71   Yes, we'll get more opportunities to optimize concat nodes:

screen shot 2018-05-15 at 9 59 28 pm

File: lib/Optimizer/GraphOptimizer.cpp (hunk 2 above)
Comment (nadavrot) [reply lib/Optimizer/GraphOptimizer.cpp]: @jfix71 It's okay for mayDepend to fail. This is a compile-time vs performance tradeoff. If we can't trace the graph to the depth of 6 then we should probably not try to analyze this region and give up. I think that in practice this won't be a real performance problem. Heuristics like this are used in LLVM and GCC.

Comment (nadavrot) [reply lib/Optimizer/GraphOptimizer.cpp]: @jfix71  Looks like depth == 6 is a good constant to use  :)

https://github.com/llvm-mirror/llvm/blob/c36d516fd6b1c105a53b3dd22843979e85ad6383/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp#L127

**Comment by nadavrot** [comment]
@jfix71 Thanks for the review. I'm updating the PR.

File: lib/Optimizer/GraphOptimizer.cpp (hunk 2 above)
Comment (jfix71) [reply lib/Optimizer/GraphOptimizer.cpp]: Ah yeah it conservatively returns true if it gets to max depth.

File: lib/Optimizer/GraphOptimizer.cpp [hunk 3]
```diff
@@ -422,6 +422,121 @@ static bool sinkCode(Function *F) {
...
+// large matmul. The large matmul is more likely to utilize the hardware. The
+// result of the big matmul is the concatenated results.
+//
+//            ____      _________        _________
+//   ----    |    |    |         |     M|  A * C  |
+// M| A  |  T| B  | * K|    C    | =    |---------|
+//   ---- ,  |    |    |         |     T|  B * C  |
+//    K       ----      ---------        ---------
+//             K            R                R
+static void mergeMatMul(Function *F) {
```
Comment (rdzhabarov) [inline lib/Optimizer/GraphOptimizer.cpp]: I think for quantized matrices it's not going to be that simple. In order to merge several matrices, we need to make sure that {S,O} for the A and B are the same.

File: lib/Optimizer/GraphOptimizer.cpp [hunk 4]
```diff
@@ -422,6 +422,121 @@ static bool sinkCode(Function *F) {
...
+      leftMatrixUsers[MM->getLHS().getNode()].push_back(MM);
+    }
+  }
+
+  // Merge RHS matrices.
+  for (auto &it : rightMatrixUsers) {
+    auto &MMs = it.second;
+
+    // Collects the LHS values to merge.
+    std::vector<NodeValue> LHS;
```
Comment (rdzhabarov) [inline lib/Optimizer/GraphOptimizer.cpp]: we only can merge same {S,O} LHS values

Comment (nadavrot) [reply lib/Optimizer/GraphOptimizer.cpp]: @rdzhabarov You are right. Good catch. I will fix this.

Comment (jfix71) [reply lib/Optimizer/GraphOptimizer.cpp]: This is a good point, however I do want to point out that this optimization occurs both pre and post quantization/profiling. I think it is most likely to apply to mat muls pre-quantization (on the very first optimization pass), correct? Which brings up an interesting point -- this mat mul merging will force mat muls that individually/previously would have their own quantization parameters to instead now all share the same parameters, meaning they will likely have worse precision at the cost of better utilization. (Of course, this is the tradeoff for quantization in the first place 😄)

**Pull request merged**